src/Build/Graph/GraphBuilder.cs

```diff
@@ -0,0 +1,397 @@
+                        task.ContinueWith(
+                            _ =>
+                            {
+                                if (task.IsFaulted)
+                                {
+                                    exceptionsInTasks.Add(task.Exception.InnerException);
+                                }
+                                tasksInProgress.TryRemove(projectToEvaluate, out var _);
+                                evaluationWaitHandle.Set();
+                            });
+                        task.Start();
+                    }
+                }
+                else
+                {
+                    // if projectsToEvaluate is empty but there are tasks in progress, there is nothing to do till a task completes and discovers new projects
+                    // wait till a task completes and sends a signal
+                    evaluationWaitHandle.WaitOne();
+                }
+            }
+
+            if (exceptionsInTasks.Count != 0)
+            {
+                exceptions = exceptionsInTasks.ToList();
+                return false;
+            }
+
+            exceptions = null;
+            return true;
+        }
+
+        internal static string FormatCircularDependencyError(List<string> projectsInCycle)
+        {
+            const int MAX_PATH = 260;
+            var errorMessage = new StringBuilder(projectsInCycle.Count * MAX_PATH);
+            errorMessage.AppendLine();
+            for (var i = projectsInCycle.Count - 1; i >= 0; i--)
+            {
+                if (i != 0)
+                {
+                    errorMessage.Append(projectsInCycle[i])
+                        .Append(" ->")
+                        .AppendLine();
+                }
+                else
+                {
+                    errorMessage.Append(projectsInCycle[i]);
+                }
+            }
+
+            return errorMessage.ToString();
+        }
+
+        private static PropertyDictionary<ProjectPropertyInstance> CreatePropertyDictionary(IDictionary<string, string> properties)
+        {
+            PropertyDictionary<ProjectPropertyInstance> propertyDictionary;
+            if (properties == null)
+            {
+                propertyDictionary = new PropertyDictionary<ProjectPropertyInstance>(0);
+            }
+            else
+            {
+                propertyDictionary = new PropertyDictionary<ProjectPropertyInstance>(properties.Count);
+                foreach (var entry in properties)
+                {
+                    propertyDictionary[entry.Key] = ProjectPropertyInstance.Create(entry.Key, entry.Value);
+                }
+            }
+
+            return propertyDictionary;
+        }
+
+        internal class GraphEdges
+        {
+            private ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance> ReferenceItems =
+                new ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance>();
```

Comment:
This data type raises some concerns for me. Can you expand on why it's needed?

Reply:
The change requires keeping extra state for each edge (the project reference metadata for each reference). There's two ways to keep it:
1. Enhance the OM by adding a GraphEdge class, and then have ProjectGraphNode reference lists of GraphEdge objects, rather than directly referencing parent/child ProjectGraphNodes
2. Hack it out by introducing a dictionary.

1. implies either changing the public API to add GraphEdge, or making it private by adding a dictionary on each node, which is memory wasteful. Since it's not clear whether anyone would find the extra public API useful, I decided to just go with 2. and keep everything private. If, in the future, somebody needs to access this data, then we can spend more energy in designing some cute looking API.